Src/Plugins/Security/SEC101_043.NuGetCredentialsValidator.cs

```diff
@@ -0,0 +1,222 @@
+                        return ReturnUnknownAuthorization(ref message, host, account: username);
+                    }
+
+                    case HttpStatusCode.Unauthorized:
+                    case HttpStatusCode.Forbidden:
+                    {
+                        // Credentials may resolve this, try again with them.
+                        byte[] byteArray = Encoding.ASCII.GetBytes($"{username}:{password}");
+                        client.DefaultRequestHeaders.Authorization =
+                            new AuthenticationHeaderValue("Basic", Convert.ToBase64String(byteArray));
+
+                        using HttpResponseMessage responseWithCredentials = client
+                            .GetAsync(host, HttpCompletionOption.ResponseHeadersRead)
+                            .GetAwaiter()
+                            .GetResult();
+
+                        switch (responseWithCredentials.StatusCode)
+                        {
+                            case HttpStatusCode.OK:
+                            {
+                                // Credentials resolved the forbidden/unauthorized message we received.
+                                return ReturnAuthorizedAccess(ref message, asset: host, account: username);
+                            }
+
+                            case HttpStatusCode.Forbidden:
+                            case HttpStatusCode.Unauthorized:
+                            {
+                                return ReturnUnauthorizedAccess(ref message, asset: host, account: username);
+                            }
+
+                            default:
+                            {
+                                message = CreateUnexpectedResponseCodeMessage(responseWithCredentials.StatusCode);
+                                return ValidationState.Unknown;
+                            }
+                        }
+                    }
+
+                    default:
+                    {
+                        message = CreateUnexpectedResponseCodeMessage(responseWithNoCredentials.StatusCode);
+                        return ValidationState.Unknown;
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                return ReturnUnhandledException(ref message, e, asset: host, account: username);
+            }
+        }
+
+        private static List<string> ExtractHostsHelper(string hostXmlAsString)
+        {
+            var hostXml = new XmlDocument();
+            hostXml.LoadXml(hostXmlAsString);
+
+            // First attempt the most common format for package sources: <packageSources><add key="..." value="{the thing we're interested in}" /></packageSources>
+            var returnList = hostXml?.ChildNodes[0] // <packageSources>...
+                            ?.ChildNodes.Cast<XmlNode>().Where(x => x.Name.Equals("add", StringComparison.OrdinalIgnoreCase)) // <add ... >  <clear/> (the first node)
```

Comment:
>< [](http://example.com/codeflow?start=45&length=1)

this code is not readable.

Reply:
changed a little bit, moved the comments above the linq query.